File: AI-powered-speech-training-fe/lib/models/recording.dart
import 'feedback.dart';

class Recording {
  final String id;
  final String topicId;
  final String topicTitle;
  final String audioUrl;
  final int duration;
  final String createdAt;
  final String transcript;
  final Feedback feedback;

  Recording({
    required this.id,
    required this.topicId,
    required this.topicTitle,
    required this.audioUrl,
    required this.duration,
    required this.createdAt,
    required this.transcript,
    required this.feedback,
  });

  factory Recording.fromJson(Map<String, dynamic> json) {
    return Recording(
      id: json['_id'] as String? ?? json['id'] as String,
      topicId: json['topicId'] is Map
          ? json['topicId']['_id'] as String
          : json['topicId'] as String,
      topicTitle: json['topicTitle'] as String? ?? '',
      audioUrl: json['audioUrl'] as String? ?? '',
      duration: (json['duration'] as num?)?.toInt() ?? 0,
      createdAt: json['createdAt'] as String? ?? '',
      transcript: json['transcript'] as String? ?? '',
      feedback: json['feedback'] != null
          ? Feedback.fromJson(json['feedback'] as Map<String, dynamic>)
          : Feedback(
              overall: 0,
              fluency: 0,
              pronunciation: 0,
              grammar: 0,
              vocabulary: 0,
              coherence: 0,
              strengths: [],
              issues: [],
              suggestions: [],
            ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'topicId': topicId,
      'topicTitle': topicTitle,
      'audioUrl': audioUrl,
      'duration': duration,
      'createdAt': createdAt,
      'transcript': transcript,
      'feedback': feedback.toJson(),
    };
  }
}

File: AI-powered-speech-training-be/models/Recording.js
const mongoose = require('mongoose');

const feedbackSchema = new mongoose.Schema({
  overall: { type: Number, min: 0, max: 10, default: 0 },
  fluency: { type: Number, min: 0, max: 10, default: 0 },
  pronunciation: { type: Number, min: 0, max: 10, default: 0 },
  grammar: { type: Number, min: 0, max: 10, default: 0 },
  vocabulary: { type: Number, min: 0, max: 10, default: 0 },
  coherence: { type: Number, min: 0, max: 10, default: 0 },
  strengths: [String],
  issues: [String],
  suggestions: [String],
}, { _id: false });

const recordingSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  topicId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Topic',
    required: true,
  },
  audioUrl: {
    type: String,
    required: true,
  },
  duration: {
    type: Number,
    default: 0,
  },
  transcript: {
    type: String,
    default: '',
  },
  feedback: {
    type: feedbackSchema,
    default: () => ({}),
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
});

recordingSchema.index({ userId: 1, createdAt: -1 });

module.exports = mongoose.model('Recording', recordingSchema);
